### PR TITLE
Add more metadata about the target if possible

### PR DESCRIPTION
**Pull request by geobeau**
Changes proposed in this PR:
- Add more metadata when creating a service in Consul. It adds the name of the node name (on kubernetes) and the references of the entries in the endpoints of the kubernetes service.

How I've tested this PR:
- manually

Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added (*HashiCorp engineers only, community PRs should not add a changelog entry*)

### CODE REVIEW

File: catalog/to-consul/resource.go [hunk 1]
```diff
@@ -643,6 +646,17 @@ func (t *ServiceResource) registerServiceInstance(
 			r.Service.ID = serviceID(r.Service.Service, addr)
 			r.Service.Address = addr
 			r.Service.Port = epPort
+			r.Service.Meta = make(map[string]string)
+			for k, v := range baseService.Meta {
+				r.Service.Meta[k] = v
+			}
```
Comment (lkysow) [inline catalog/to-consul/resource.go:652]: Consul metadata has some specific rules that must be followed so we can't map directly:

> The meta object is a map of max 64 key/values with string semantics. Key can contain only ASCII chars and no special characters (A-Z a-z 0-9 _ and -). For performance and security reasons, values as well as keys are limited to 128 characters for keys, 512 for values. This object has the same limitations as the node meta object in node definition. All those meta data can be retrieved individually per instance of the service and all the instances of a given service have their own copy of it.

From https://www.consul.io/docs/discovery/services

**Review by lkysow — COMMENTED** [review]
Thanks for the PR! Can you write tests for the changes as well please.

**Comment by geobeau** [comment]
@lkysow updated the PR with your comments. Sorry for the ~slight~ delay 😄

**Comment by lkysow** [comment]
@geobeau would you be able to rebase off of `main`? I'm not sure why our github checks aren't running.

**Comment by geobeau** [comment]
Done

**Comment by geobeau** [comment]
Maybe I can try to close the PR and repoen it if it still doesn't work

File: control-plane/catalog/to-consul/resource_test.go [hunk 2]
```diff
@@ -1266,6 +1266,39 @@ func TestServiceResource_clusterIPTargetPortNamed(t *testing.T) {
...
+	// Verify what we got
+	retry.Run(t, func(r *retry.R) {
+		syncer.Lock()
+		defer syncer.Unlock()
+		actual := syncer.Registrations
+		require.Len(r, actual, 2)
+		require.Equal(r, "foobar", actual[0].Service.Meta[ConsulK8SRefValue])
+		require.Equal(r, "pod", actual[0].Service.Meta[ConsulK8SRefKind])
+		require.NotContains(r, ConsulK8SRefValue, actual[0].Service.Meta)
+		require.NotContains(r, ConsulK8SRefKind, actual[0].Service.Meta)
```
Comment (lkysow) [inline control-plane/catalog/to-consul/resource_test.go:1298]: What is this testing?

Comment (lkysow) [inline control-plane/catalog/to-consul/resource_test.go:1298]: What is this testing?

File: control-plane/catalog/to-consul/resource_test.go [hunk 3]
```diff
@@ -1266,6 +1266,39 @@ func TestServiceResource_clusterIPTargetPortNamed(t *testing.T) {
...
+	})
```
Comment (lkysow) [inline control-plane/catalog/to-consul/resource_test.go:1299]: Can we also test the node name meta?

Comment (lkysow) [inline control-plane/catalog/to-consul/resource_test.go:1299]: Can we also test the node name meta?

**Review by lkysow — COMMENTED** [review]
Looks pretty good. Don't worry about the tests, we need to fix them.

File: control-plane/catalog/to-consul/resource_test.go (hunk 3 above)
Comment (geobeau) [reply control-plane/catalog/to-consul/resource_test.go:1303]: Done

File: control-plane/catalog/to-consul/resource_test.go [hunk 4]
```diff
@@ -1612,7 +1646,7 @@ func createEndpoints(t *testing.T, client *fake.Clientset, serviceName string, n
 			Subsets: []apiv1.EndpointSubset{
 				{
 					Addresses: []apiv1.EndpointAddress{
-						{NodeName: &node1, IP: "1.1.1.1"},
+						{NodeName: &node1, IP: "1.1.1.1", TargetRef: &target_ref},
```
[the remark this answers is not in the document]
Comment (geobeau) [reply control-plane/catalog/to-consul/resource_test.go]: Done

File: control-plane/catalog/to-consul/resource_test.go (hunk 2 above)
Comment (geobeau) [reply control-plane/catalog/to-consul/resource_test.go]: Actually it was a mistake, I redid it and added a comment. It was supposed to test that the second service didn't contain incorrect refs

**Review by lkysow — APPROVED** [review]

**Pull request merged**